packages/console/src/pages/SignInExperience/components/ManageLanguageModal/AddLanguageSelector.tsx

```diff
@@ -12,47 +15,87 @@ type Props = {
+  const { t } = useTranslation(undefined, { keyPrefix: 'admin_console' });
+  const selectorRef = useRef<HTMLDivElement>(null);
+  const searchInputRef = useRef<HTMLInputElement>(null);
   const [isDropDownOpen, setIsDropDownOpen] = useState(false);
+  const [searchInputValue, setSearchInputValue] = useState('');
+
+  const filteredOptions = searchInputValue
+    ? options.filter(
+        (languageTag) =>
+          languageTag.toLocaleLowerCase().includes(searchInputValue.toLocaleLowerCase()) ||
+          uiLanguageNameMapping[languageTag]
+            .toLocaleLowerCase()
+            .includes(searchInputValue.toLocaleLowerCase())
+      )
+    : options;
+
+  const clickOutsideHandler = ({ target }: MouseEvent) => {
+    if (target instanceof HTMLElement && !selectorRef.current?.contains(target)) {
+      setIsDropDownOpen(false);
+      setSearchInputValue('');
+    }
+  };
+
+  useEffect(() => {
+    if (isDropDownOpen) {
+      searchInputRef.current?.focus();
+      document.addEventListener('mousedown', clickOutsideHandler);
+    } else {
+      document.removeEventListener('mousedown', clickOutsideHandler);
+    }
+
+    return () => {
+      if (isDropDownOpen) {
+        document.removeEventListener('mousedown', clickOutsideHandler);
+      }
+    };
+  }, [isDropDownOpen, searchInputRef]);
 
   return (
-    <div>
-      <div ref={anchorRef}>
+    <div ref={selectorRef} className={style.languageSelector}>
+      <div className={style.input}>
         <Button
-          className={style.addLanguageButton}
-          icon={<Plus className={style.iconPlus} />}
+          className={classNames(style.addLanguageButton, isDropDownOpen && style.hidden)}
+          icon={<Plus className={style.buttonIcon} />}
           title="sign_in_exp.others.manage_language.add_language"
           type="outline"
           size="medium"
           onClick={() => {
             setIsDropDownOpen(true);
           }}
         />
+        <TextInput
+          ref={searchInputRef}
+          icon={<SearchIcon className={style.buttonIcon} />}
+          className={classNames(!isDropDownOpen && style.hidden)}
+          placeholder={t('general.type_to_search')}
+          value={searchInputValue}
+          onChange={(event: ChangeEvent<HTMLInputElement>) => {
+            setSearchInputValue(event.target.value);
+          }}
+        />
       </div>
-      <Dropdown
-        isFullWidth
-        anchorRef={anchorRef}
-        isOpen={isDropDownOpen}
-        onClose={() => {
-          setIsDropDownOpen(false);
-        }}
-      >
-        {options.map((languageTag) => (
-          <DropdownItem
-            key={languageTag}
-            onClick={() => {
-              onSelect(languageTag);
-            }}
-          >
-            <div className={style.dropDownItem}>
-              <div className={style.languageName}>{languages[languageTag]}</div>
+      {isDropDownOpen && filteredOptions.length > 0 && (
```

Comment:
How about using a css selector (focus) to control the open/close state?

Reply:
Tried, but I found it complicated to control states between 3 elements (the add button, search input, and drop-down list) by CSS, and also the existing `TextInput` component with an icon is not perfect to handle the focus state, so add an issue to add searching functionality to `Select` or `DropDown` component: [LOG-4340: Support Searching In `<Select />` or `<DropDown />` Component](https://linear.app/silverhand/issue/LOG-4340/support-searching-in-select-or-dropdown-component).
Currently, we decide to move forward with this implementation.
@wangsijie